[Security Solution] Installation review pagination: Backend

nikitaindik opened the pull request:
**Partially addresses: #241656**

> **Note:** In this PR the endpoint is paginated, but `page` and `per_page` parameters are optional. If no pagination parameters are provided, it will return all rules at once (same as previous behavior). We want to release this PR first and then we'll make the pagination parameters required in a follow-up PR.

## Summary

This PR adds pagination, sorting and filtering functionality to the prebuilt rules installation review endpoint (`internal/prebuilt_rules/installation/_review`) in accordance with the [API design proposal](https://docs.google.com/document/d/1AYlt8wJMoLD-V_owAd4qL-h76IOVnfDCZ77VNmmf-Ks/edit?usp=sharing) (internal).

No changes to the front end in this PR. I will make frontend use backend-side pagination, sorting and filtering in a follow-up PR.

## Changes
- Implemented pagination, sorting and filtering of prebuilt rules. `page` and `per_page` parameters are optional. If these parameters are not provided, endpoint will return all rules at once (same as previous behaviour).
- Added mappings for more fields of `security-rule` SO: `name`, `tags`, `severity`, `risk_score`
- Split Prebuilt Rule Assets client into multiple files, because it became to large
- Updated existing Prebuilt Rule Assets client methods to use `savedObjectsClient.search` instead of `savedObjectsClient.find`. This allows to use the "terms" aggregation to fetch needed rules without having to work around the "max boolean clauses count reached" issue.
- Covered the changes with integration tests according to the recently merged test plan ([PR](https://github.com/elastic/kibana/pull/244804))
- Added a couple of new scenarios to test plans based on [suggestions](https://github.com/elastic/kibana/pull/244804#issuecomment-3634810653) from @pborgonovi



<!--ONMERGE {"backportTargets":["8.19","9.2","9.3"]} ONMERGE-->

nikitaindik committed:
Backend changes

nikitaindik committed:
Revert concurrency limit back to 1

nikitaindik committed:
Make parameters optional

nikitaindik committed:
Update schemas in SO registration

kibanamachine committed:
Changes from node scripts/jest_integration -u src/core/server/integration_tests/ci_checks

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/376666)
* Commit: 91f733bd549ab453b20a81693f333867267275b3

### Failed CI Steps
* [Checks](https://buildkite.com/elastic/kibana-pull-request/builds/376666#019b4ff9-a88f-4f30-b536-672c22f0a9e0)

### History
* :broken_heart: [Build #376466](https://buildkite.com/elastic/kibana-pull-request/builds/376466) failed b6651e12d450d03f319bf3f2c21ba2a7620181a5

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376666","number":376666,"commit":"91f733bd549ab453b20a81693f333867267275b3"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376466","number":376466,"commit":"b6651e12d450d03f319bf3f2c21ba2a7620181a5"}],"number":376666}
buildkite-pr-comment-->

nikitaindik committed:
Improve defaulting of request parameters

nikitaindik committed:
Add a fixture for mappings

nikitaindik committed:
Remove unnecessary parameters in tests

nikitaindik committed:
Extract methods of prebuilt rule assets client into separate files

nikitaindik committed:
Remove the "create" validation from schema

kibanamachine committed:
Changes from node scripts/jest_integration -u src/core/server/integration_tests/ci_checks

nikitaindik committed:
Refactor prebuilt rule assets client

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/376764)
* Commit: a50b292bf9ca5bf9d9ff88b757107552ce032437

### Failed CI Steps
* [Check changes in Saved Objects](https://buildkite.com/elastic/kibana-pull-request/builds/376764#019b55b7-875c-47af-94ed-3b7fc7f311e4)

### History
* :broken_heart: [Build #376755](https://buildkite.com/elastic/kibana-pull-request/builds/376755) failed 1ee0e1de44a31d12fcac06df69039b4bbc1e31bd
* :broken_heart: [Build #376733](https://buildkite.com/elastic/kibana-pull-request/builds/376733) failed fac56bba697e1a448bb66b1ff765160ae9d3729f
* :broken_heart: [Build #376696](https://buildkite.com/elastic/kibana-pull-request/builds/376696) failed 658c51ec833c2eaca178230f2590cffefb061a70
* :broken_heart: [Build #376675](https://buildkite.com/elastic/kibana-pull-request/builds/376675) failed 056f59d92ab8c9232d0df3261161fa4a01cc1781
* :broken_heart: [Build #376666](https://buildkite.com/elastic/kibana-pull-request/builds/376666) failed 91f733bd549ab453b20a81693f333867267275b3

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376764","number":376764,"commit":"a50b292bf9ca5bf9d9ff88b757107552ce032437"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376755","number":376755,"commit":"1ee0e1de44a31d12fcac06df69039b4bbc1e31bd"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376733","number":376733,"commit":"fac56bba697e1a448bb66b1ff765160ae9d3729f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376696","number":376696,"commit":"658c51ec833c2eaca178230f2590cffefb061a70"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376675","number":376675,"commit":"056f59d92ab8c9232d0df3261161fa4a01cc1781"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376666","number":376666,"commit":"91f733bd549ab453b20a81693f333867267275b3"}],"number":376764}
buildkite-pr-comment-->

nikitaindik committed:
Add logging

nikitaindik committed:
Ignore `forwardCompatibility` and `create` validation

kibanamachine committed:
Changes from node scripts/jest_integration -u src/core/server/integration_tests/ci_checks

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/376814)
* Commit: 8ada89d9170499ffa5bd62ceb9992d4e7f935369

### Failed CI Steps
* [Check changes in Saved Objects](https://buildkite.com/elastic/kibana-pull-request/builds/376814#019b5ae8-b2a9-4794-8080-87c714de6a42)

### History
* :broken_heart: [Build #376809](https://buildkite.com/elastic/kibana-pull-request/builds/376809) failed b2aa07087c9b608e77b1cd815f50a2184c5534d6
* :broken_heart: [Build #376772](https://buildkite.com/elastic/kibana-pull-request/builds/376772) failed e178c9ed8569fa5373b49a3428a452f17eab006e
* :broken_heart: [Build #376764](https://buildkite.com/elastic/kibana-pull-request/builds/376764) failed a50b292bf9ca5bf9d9ff88b757107552ce032437
* :broken_heart: [Build #376755](https://buildkite.com/elastic/kibana-pull-request/builds/376755) failed 1ee0e1de44a31d12fcac06df69039b4bbc1e31bd
* :broken_heart: [Build #376733](https://buildkite.com/elastic/kibana-pull-request/builds/376733) failed fac56bba697e1a448bb66b1ff765160ae9d3729f

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376814","number":376814,"commit":"8ada89d9170499ffa5bd62ceb9992d4e7f935369"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376809","number":376809,"commit":"b2aa07087c9b608e77b1cd815f50a2184c5534d6"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376772","number":376772,"commit":"e178c9ed8569fa5373b49a3428a452f17eab006e"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376764","number":376764,"commit":"a50b292bf9ca5bf9d9ff88b757107552ce032437"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376755","number":376755,"commit":"1ee0e1de44a31d12fcac06df69039b4bbc1e31bd"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/376733","number":376733,"commit":"fac56bba697e1a448bb66b1ff765160ae9d3729f"}],"number":376814}
buildkite-pr-comment-->

nikitaindik committed:
Update tests

maximpn commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/prebuilt_rule_installation.md:
```diff
@@ -757,6 +776,39 @@ Examples:
+Given <num_prebuilt_rule_assets> non-installed prebuilt rule assets exist in Kibana
+When user calls the review installation endpoint with <page> and <per_page> parameters
+And <page> * <per_page> is greater than N
+Then the endpoint should return an empty list of rules
```
```suggestion
Given <num_prebuilt_rule_assets> non-installed prebuilt rule assets exist in Kibana
When user calls the review installation endpoint with <page> and <per_page> parameters
And <page> * <per_page> is greater than <num_prebuilt_rule_assets>
Then the endpoint should return an empty list
```

maximpn commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/review_rule_installation/review_rule_installation_handler.ts:
```diff
@@ -34,48 +64,100 @@ export const reviewRuleInstallationHandler = async (
     const mlAuthz = ctx.securitySolution.getMlAuthz();
 
-    const allLatestVersions = await ruleAssetsClient.fetchLatestVersions();
     const currentRuleVersions = await ruleObjectsClient.fetchInstalledRuleVersions();
```
Let's rename `currentRuleVersions` to `installedRuleVersions`.

maximpn commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_client/methods/fetch_assets_by_version.ts:
```diff
@@ -0,0 +1,74 @@
+    }
+  >({
+    type: PREBUILT_RULE_ASSETS_SO_TYPE,
+    namespaces: ['default'],
```
Have you checked it's safe to use the `default` namespace? Shall we use the current namespace instead?

My concern is that the current user may not have permissions to access the `default` namespace and the SO client may restrict SOs access in that case.

maximpn commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_client/methods/fetch_tags_by_version.ts:
```diff
@@ -0,0 +1,58 @@
+      unique_tags: {
+        terms: {
+          field: `${PREBUILT_RULE_ASSETS_SO_TYPE}.tags`,
+          size: 10_000,
```
nit: Could we avoid magic constants? Ideally we should use a well known constant. This constant should have a good description why 10K is the number.

maximpn commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_client/utils.ts:
```diff
@@ -0,0 +1,60 @@
+  return queryFilter;
+}
+
+export function buildEsQuerySort(sort?: PrebuiltRuleAssetsSort): Sort | undefined {
```
```suggestion
export function prapreEsQueryDslSort(sort?: PrebuiltRuleAssetsSort): Sort | undefined {
```

maximpn commented on x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/install_prebuilt_rules/review_installation.ts:
```diff
@@ -0,0 +1,844 @@
+      });
+    });
+
+    describe('Using endpoint without providing all parameters', () => {
```
```suggestion
    describe('parameters defaults', () => {
```

maximpn commented on x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/install_prebuilt_rules/review_installation.ts:
```diff
@@ -0,0 +1,844 @@
+        });
+
+        expect(response).toMatchObject({
+          rules: [expect.objectContaining({ rule_id: 'rule-3', version: 1 })],
```
Shall we assert only the response list length?

maximpn commented on x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/install_prebuilt_rules/review_installation.ts:
```diff
@@ -0,0 +1,844 @@
+          filter: { fields: { tags: { include: ['tag-b'] } } },
+        });
+
+        expect(response.rules).toEqual([
```
It should be enough to assert only `rule_id`s.

maximpn commented on x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/install_prebuilt_rules/review_installation.ts:
```diff
@@ -0,0 +1,844 @@
+            sort: [{ field: 'name', order: 'asc' }],
+          });
+
+          expect(ascSortResponse.rules).toEqual([
```
It should be enough to assert only `rule_id`s.

nikitaindik replied:
But I also need to make sure that the latest version of the rule is used for sorting.

nikitaindik replied:
Please disregard two previous comments. I simplified assertions and made a separate test for using the latest version in sorting.

maximpn commented on x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/install_prebuilt_rules/review_installation.ts:
```diff
@@ -0,0 +1,844 @@
+            sort: [{ field: 'name', order: 'desc' }],
+          });
+
+          expect(descSortResponse.rules).toEqual([
```
It should be enough to assert only `rule_id`s.

maximpn commented on x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/install_prebuilt_rules/review_installation.ts:
```diff
@@ -0,0 +1,844 @@
+              name: 'Rule 2, previous version',
+            }),
+            createRuleAssetSavedObject({ rule_id: 'rule-2', version: 2, name: 'Rule 2' }),
+          ];
```
It's better to have the setup logic in the each test. We'll have duplications this way but it will help avoid jumping when reading the tests. On of that you may use test scenario focused data in each case.

maximpn reviewed (changes requested):
@nikitaindik Thanks for making improvements in the prebuilt rules installation review API endpoint 👍 

I left some comments regarding the implementation. There are some open questions we should clarify before merging this PR. On top of that the implementation could be simplified a bit so I left comments regarding this part as well.

banderror commented:
@nikitaindik Going to review when @maximpn's comments are addressed and CI is green

nikitaindik committed:
Add a test for custom space users

nikitaindik commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/review_rule_installation/review_rule_installation_handler.ts:
```diff
@@ -34,48 +64,100 @@ export const reviewRuleInstallationHandler = async (
+      ? await getInstallableRuleVersions(ruleAssetsClient, logger, mlAuthz, currentRuleVersionsMap)
+      : installableVersions;
+
+    const installableVersionsPage = installableVersions.slice((page - 1) * perPage, page * perPage);
```
I think we can't do it in a simple way, because we need to paginate through installable assets only (latest assets minus already installed). So we first need to determine which assets are installable and then take a page from this subset. As an alternative, we could pass a list of installable asset IDs to the terms query, so we could use `from` and `size`, but this seems more complicated and will require passing a potentially huge list of IDs in a request. Tell me if I misunderstood your idea.

nikitaindik commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_client/methods/fetch_latest_assets.ts:
```diff
@@ -0,0 +1,67 @@
+import { invariant } from '../../../../../../../../common/utils/invariant';
+
+/**
+ * Fetches the latest version of each prebuilt rule asset.
```
Makes sense!

nikitaindik commented on x-pack/solutions/security/test/security_solution_api_integration/test_suites/detections_response/rules_management/prebuilt_rules/common/install_prebuilt_rules/review_installation.ts:
```diff
@@ -0,0 +1,844 @@
+            sort: [{ field: 'risk_score', order: 'desc' }],
+          });
+
+          expect(descSortResponse.rules).toEqual([
```
Updated the test to only assert risk score.

nikitaindik committed:
Address code review comments

nikitaindik committed:
Uncomment commented-out tests

nikitaindik committed:
Update SO definition and apply validation changes from the Core team

kibanamachine committed:
Changes from node scripts/jest_integration -u src/core/server/integration_tests/ci_checks

nikitaindik commented on packages/kbn-check-saved-objects-cli/src/snapshots/validate_changes.ts:
```diff
@@ -37,11 +38,38 @@ export function validateChangesExistingType({ from, to }: ValidateChangesExistin
-    throw new Error(
-      `❌ The SO type '${name}' is defining two (or more) new model versions. Please refer to our troubleshooting guide: https://www.elastic.co/docs/extend/kibana/saved-objects#troubleshooting`
-    );
+  if (
```
This bit needs to be reviewed by the Core team

nikitaindik commented:
> Going to review when @maximpn's comments are addressed and CI is green

@banderror I addressed the comments and made CI pass. Can be reviewed now.

banderror commented:
@nikitaindik Please resolve all the comments that have been addressed

jloleysens committed:
updated SO definition

jloleysens reviewed (approved):
Apologies for the delay on this. I've updated the PR with the latest changes in main so only your original SO changes should remain.

Did not test the rest of the PR again (good idea to just double check locally!!), rest of Core changes LGTM.

kibanamachine committed:
Changes from node scripts/jest_integration -u src/core/server/integration_tests/ci_checks

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/379052)
* Commit: f0881160f168e55b1f78eaca7531f71989184200

### Failed CI Steps
* [Check changes in Saved Objects](https://buildkite.com/elastic/kibana-pull-request/builds/379052#019b9d65-8747-4e14-993e-7be0312a89ac)

### History
* :broken_heart: [Build #379031](https://buildkite.com/elastic/kibana-pull-request/builds/379031) failed 3dad6112e9c129cce393e2e9e46193182bd6f489

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/379052","number":379052,"commit":"f0881160f168e55b1f78eaca7531f71989184200"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/379031","number":379031,"commit":"3dad6112e9c129cce393e2e9e46193182bd6f489"}],"number":379052}
buildkite-pr-comment-->

maximpn reviewed (commented):
I left extra a couple of comments. We synced with @nikitaindik via Zoom to discuss the open comments regarding the review installation API endpoint implementation. In result we agreed that it'd be good to extract stats fetching and calculation into a separate function for better maintainability.

nikitaindik committed:
Address additional code review feedback

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/380134)
* Commit: e4081a99eb45cd292a0be38a5d48405f25ac5202

### Failed CI Steps
* [Check changes in Saved Objects](https://buildkite.com/elastic/kibana-pull-request/builds/380134#019bb1fa-3486-4122-b048-7307b57e3b28)

### History
* :broken_heart: [Build #379888](https://buildkite.com/elastic/kibana-pull-request/builds/379888) failed 6188738545e0ccfef18621fe5c34b149b71d6ed3
* :broken_heart: [Build #379052](https://buildkite.com/elastic/kibana-pull-request/builds/379052) failed f0881160f168e55b1f78eaca7531f71989184200
* :broken_heart: [Build #379031](https://buildkite.com/elastic/kibana-pull-request/builds/379031) failed 3dad6112e9c129cce393e2e9e46193182bd6f489

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/380134","number":380134,"commit":"e4081a99eb45cd292a0be38a5d48405f25ac5202"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/379888","number":379888,"commit":"6188738545e0ccfef18621fe5c34b149b71d6ed3"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/379052","number":379052,"commit":"f0881160f168e55b1f78eaca7531f71989184200"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/379031","number":379031,"commit":"3dad6112e9c129cce393e2e9e46193182bd6f489"}],"number":380134}
buildkite-pr-comment-->

maximpn reviewed (approved):
@nikitaindik Thanks for addressing my comments 🙏 

I left a comment regarding the function declaration and disabling a linting rule. Besides that the PR LGTM.

Local testing hasn't revealed issues.

Approving in advance. Please address my comment before merging.

nikitaindik committed:
Small adjustments after code review

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/380819)
* Commit: 5655f2842e9c2e63473c4191b995a6e04d50bc99

### Failed CI Steps
* [Check changes in Saved Objects](https://buildkite.com/elastic/kibana-pull-request/builds/380819#019bb6ef-d981-4751-b540-e3079e7537df)

### History
* :broken_heart: [Build #380134](https://buildkite.com/elastic/kibana-pull-request/builds/380134) failed e4081a99eb45cd292a0be38a5d48405f25ac5202
* :broken_heart: [Build #379888](https://buildkite.com/elastic/kibana-pull-request/builds/379888) failed 6188738545e0ccfef18621fe5c34b149b71d6ed3
* :broken_heart: [Build #379052](https://buildkite.com/elastic/kibana-pull-request/builds/379052) failed f0881160f168e55b1f78eaca7531f71989184200
* :broken_heart: [Build #379031](https://buildkite.com/elastic/kibana-pull-request/builds/379031) failed 3dad6112e9c129cce393e2e9e46193182bd6f489

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/380819","number":380819,"commit":"5655f2842e9c2e63473c4191b995a6e04d50bc99"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/380134","number":380134,"commit":"e4081a99eb45cd292a0be38a5d48405f25ac5202"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/379888","number":379888,"commit":"6188738545e0ccfef18621fe5c34b149b71d6ed3"},{"buildStatus":{"state":"passed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/379052","number":379052,"commit":"f0881160f168e55b1f78eaca7531f71989184200"},{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/379031","number":379031,"commit":"3dad6112e9c129cce393e2e9e46193182bd6f489"}],"number":380819}
buildkite-pr-comment-->

gsoldevila committed:
Update schemas to include mapped types

nikitaindik committed:
Merge

nikitaindik committed:
Add fixture for 10.2.0

banderror commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/review_rule_installation/review_rule_installation_handler.ts:
```diff
@@ -34,48 +64,113 @@ export const reviewRuleInstallationHandler = async (
+      installableVersionsPage
     );
 
+    const { tags, numRulesToInstall } = await fetchStats();
```
Stats can be calculated in parallel with the rules fetching logic. It would likely make the endpoint faster.

banderror reviewed (commented):
Smoke tested locally and reviewed code changes. The PR LGTM, and the comments I left are mostly minor.

kibanamachine committed:
Changes from node scripts/jest_integration -u src/core/server/integration_tests/ci_checks

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/381914)
* Commit: 83ce02b801826cf7183cbab8398f4226602ddba5

### Failed CI Steps
* [Checks](https://buildkite.com/elastic/kibana-pull-request/builds/381914#019bc114-9bbe-4647-8594-ed133a0deefc)

### History
* :broken_heart: [Build #381637](https://buildkite.com/elastic/kibana-pull-request/builds/381637) failed a10460ee3bd95e08e5e8f0f3028a1d246058b040
* :broken_heart: [Build #381533](https://buildkite.com/elastic/kibana-pull-request/builds/381533) failed a10460ee3bd95e08e5e8f0f3028a1d246058b040
* :broken_heart: [Build #381500](https://buildkite.com/elastic/kibana-pull-request/builds/381500) failed a10460ee3bd95e08e5e8f0f3028a1d246058b040
* :broken_heart: [Build #381418](https://buildkite.com/elastic/kibana-pull-request/builds/381418) failed 2b7b4a57f7c5f19abc7aaefed30fee058ee175c2
* :broken_heart: [Build #381416](https://buildkite.com/elastic/kibana-pull-request/builds/381416) failed 34b594bf4617e1cde6132e7b64d59b0b1524b13f

cc @nikitaindik

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/381914","number":381914,"commit":"83ce02b801826cf7183cbab8398f4226602ddba5"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/381637","number":381637,"commit":"a10460ee3bd95e08e5e8f0f3028a1d246058b040"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/381533","number":381533,"commit":"a10460ee3bd95e08e5e8f0f3028a1d246058b040"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/381500","number":381500,"commit":"a10460ee3bd95e08e5e8f0f3028a1d246058b040"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/381418","number":381418,"commit":"2b7b4a57f7c5f19abc7aaefed30fee058ee175c2"},{"buildStatus":{"state":"running","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/381416","number":381416,"commit":"34b594bf4617e1cde6132e7b64d59b0b1524b13f"}],"number":381914}
buildkite-pr-comment-->

nikitaindik commented on x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/prebuilt_rules/review_rule_installation/review_rule_installation_route.ts:
```diff
@@ -5,19 +5,55 @@
+    /**
+     * Rules per page
+     */
+    per_page: z.coerce.number().int().min(1).max(10_000).optional(),
```
No need for it, indeed. Removed in the [frontend PR](https://github.com/elastic/kibana/pull/248259/files).

nikitaindik commented on x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/review_rule_installation/review_rule_installation_handler.ts:
```diff
@@ -34,48 +64,113 @@ export const reviewRuleInstallationHandler = async (
+      installableVersionsPage
     );
 
+    const { tags, numRulesToInstall } = await fetchStats();
```
Refactored this in the [FE PR](https://github.com/elastic/kibana/pull/248259/files)
